server.js

```diff
@@ -1,11 +1,27 @@
-/**
- * Created by adam on 25/12/2016.
- */
+var _ = require('lodash');
 var express = require('express');
 var app = express();
+var env = require('dotenv').load();
+var envConfig = require('./lookenv.config');
+
 
 app.use(express.static('.'));
 
-app.listen(8000, function () {
+app.get('/config', function(request, response) {
```

Comment:
As it's a JSON, I think having the route as `/config.json` is more explicit.

Instead of simply being a dump generator of a JSON with the environment keys, why not generate the JSON with the actual config you want? Like:

```javascript
var config = {
    conductorUrl: envConfig.OS_CONDUCTOR_URL || envConfig.OS_BASE_URL,
    explorerUrl: envConfig.OS_EXPLORER_URL || envConfig.OS_BASE_URL,
    // ...
}
```

Basically, this route returns the config.json as we want it to be, so any place that wants a config can use it (including ourselves, if we want to debug things)

Reply:
The main advantage I see in having this logic here, and not on the frontend, is that it's easier to test as needed. Not that I think we should test it all, but it might be useful for some parts of the logic, like the `OS_SNIPPETS_`.

Reply:
A different approach I've used a couple of times is to embed the configuration inside a script element in index.html:
```html
<script>
  window.globalConfig = {{ config || safe }};
</script>
```
And then let node render index.html with the configuration inside.

This allows us to remove the `/config` endpoint, and, more importantly, simplifies the frontend app which may rely on the configuration always being there instead of having to asynchronously fetch it  (and everything else initializing asynchronously after it etc.)

An example of this approach can be seen here: https://github.com/OpenBudget/budgetkey-app-generic-item/blob/master/index.js
Where node is prefetching some data and embedding a theme configuration in index.html before serving it.

Reply:
Adding to that idea, we could instead have a `/config.js` route, similar to what we had with:

```js
window.globalConfig = {
  // ...
};
```

and then just add `<script src="config.js">`. This needs to be loaded before any other script, and we might have problems with busting the browser's cache, but then we can still use our current `app.use(express.static('.'));`.

Reply:
I wanted to avoid having `window.globalConfig` defined outside of the responsibility of the Angular app. Defining it in the initialisation code centralises it with other global definitions. It makes the Angular app responsible for its globals, and doesn't presume it will just be there.

However, we can improve the current code by removing the chained `thens` after requesting the config, and just make the dependent script requests within the first `then`.

```javascript
  // Make config global available to app, then load scripts that require
  // globalConfig.
  $.get('config.json')
    .then(function(config) {
      globals.globalConfig = Object.assign({}, globals.globalConfig, config);

      // Load snippets (requires globalConfig)
      $.getScript('public/scripts/snippets.js')
        .fail(function(jqxhr, settings, exception) {
          throw(exception);
        });

      // Load externally hosted authClient.services lib (makes `authenticate`
      // and `authorize` services available to angular modules).
      var libUrl = globalConfig.conductorUrl + '/user/lib';
      $.getScript(libUrl)
        .fail(function(jqxhr, settings, exception) {
          console.log('Unable to load authClient.services from ' + libUrl);
        });

      return globals.globalConfig;
    });
```

Reply:
Although having Angular take care of everything it needs is a good goal, I think these config variables can be treated differently. It's similar to environment variables: the code doesn't deal with how to load the env variables, they're just there, injected by someone else. The project is concerned with the code, but not the environment.

The advantage of dealing it this way is that it simplifies our code. For example, the current code doesn't deal with the case on when the `/config.json` path fails. If the config is instead injected via a `<script src="config.js">`, the `index.js` could contain code like:

```js
(function(globals) {
  if (globals.globalConfig === undefined) {
    throw "Missing globalConfig object in the global scope.";
  }
  // ...
})(window);
```

In other words, the `index.js` just ensures that the config was loaded, just to be a bit defensive.

Reply:
Having added these scripts via `$.getScript` in explorer now, I think @vitorbaptista is right. It's less error-prone and clearer to add `config.js` and `snippets.js` via normal script tags. While I don't like script relying on magic globals being present, it simplifies the initialisation process, as @akariv suggests. This also helps resolve a related issues, specific to os-explorer, where the `/config.json` path is being redirected to the legacy api in the ngnix location config.

Reply:
Thanks both for your comments and suggests here!